### PR TITLE
(feat): Signup API for email auth

### PR DESCRIPTION
**Pull request by Sanjay1611**
Signed-off-by: Sanjay Nathani <sanjay.nathani@mayadata.io>

This PR consists of following changes -

- Implement signup endpoint as `/v1/signup`
- Change boarding states based on the signup flow
- Introduce `emailmanager` in between server 

The Request-Response will be looking like this
Screenshot from 2021-03-11 01-33-32

The user will be looking like this
```
    {
        "_id": "604928a64eecb313e3d18793",
        "uid": "7b02d6f7-f82a-43c9-817e-98cd9a6fce13",
        "username": "sanjay.nathani@mayadata.io",
        "email": null,
        "unverified_email": "sanjay.nathani@mayadata.io",
        "name": "Sanjay Nathani",
        "kind": "local",
        "role": "user",
        "logged_in": true,
        "created_at": "2021-03-10T20:14:30.722Z",
        "updated_at": "2021-03-10T20:14:44.257Z",
        "removed_at": null,
        "state": "",
        "onboarding_state": 1
    }
```

### CODE REVIEW

File: manager/usermanager/createUser.go [hunk 1]
```diff
@@ -26,11 +28,28 @@ func CreateUser(userStore *store.UserStore, user *models.UserCredentials) (*mode
 	}
 
 	password := string(hashedPassword)
-	user.Password = &password
 	uid := uuid.Must(uuid.NewRandom()).String()
-	user.UID = &uid
-	err = userStore.Set(user)
-	return user.GetPublicInfo(), err
+	var role models.Role
+	if user.GetRole() != "" {
```
Comment (avishnu) [inline manager/usermanager/createUser.go:33]: What does this mean?

File: manager/usermanager/createUser.go [hunk 2]
```diff
@@ -26,11 +28,28 @@ func CreateUser(userStore *store.UserStore, user *models.UserCredentials) (*mode
...
+		role = user.Role
+	} else {
+		role = models.RoleUser
+	}
+
+	newUser := &models.UserCredentials{
```
Comment (avishnu) [inline manager/usermanager/createUser.go:39]: Why this change?

File: manager/usermanager/updateUser.go [hunk 3]
```diff
@@ -12,9 +12,27 @@ import (
 
 // UpdateUserDetails get the user information
```
Comment (avishnu) [inline manager/usermanager/updateUser.go:13]: get or set?

File: manager/usermanager/updateUser.go [hunk 4]
```diff
@@ -12,9 +12,27 @@ import (
...
 func UpdateUserDetails(userStore *store.UserStore, user *models.UserCredentials) (*models.PublicUserInfo, error) {
-	// This is just for github auth. Will implement a switch  case here once we have a self signup
-	if user.OnBoardingState == models.BoardingStateEmailVerified && user.Kind == models.GithubAuth {
-		user.OnBoardingState = models.BoardingStateVerifiedAndComplete
+	switch user.OnBoardingState {
+	case models.BoardingStateSignup:
+		{
+			if user.Email != nil {
+				user.OnBoardingState = models.BoardingStateEmailVerified
```
Comment (avishnu) [inline manager/usermanager/updateUser.go:19]: You are changing the onboarding state within the switch clause, is that fine?

File: manager/usermanager/updateUser.go [hunk 5]
```diff
@@ -12,9 +12,27 @@ import (
 
 // UpdateUserDetails get the user information
 func UpdateUserDetails(userStore *store.UserStore, user *models.UserCredentials) (*models.PublicUserInfo, error) {
-	// This is just for github auth. Will implement a switch  case here once we have a self signup
-	if user.OnBoardingState == models.BoardingStateEmailVerified && user.Kind == models.GithubAuth {
-		user.OnBoardingState = models.BoardingStateVerifiedAndComplete
+	switch user.OnBoardingState {
```
Comment (avishnu) [inline manager/usermanager/updateUser.go:15]: Please add code comments explaing the below logic for each switch case

File: pkg/models/user.go [hunk 6]
```diff
@@ -79,13 +79,13 @@ var adminUID = uuid.Must(uuid.NewRandom()).String()
...
-	UID:      &adminUID,
-	Name:     &types.DefaultName,
-	Email:    &types.DefaultEmail,
-	UserName: &types.DefaultUserName,
-	Password: &types.DefaultUserPassword,
-	Role:     RoleAdmin,
-	Kind:     LocalAuth,
+	UID:             &adminUID,
+	Name:            &types.DefaultName,
+	UnverifiedEmail: &types.DefaultUserName,
```
Comment (avishnu) [inline pkg/models/user.go:84]: If we keep this empty, will it cause any issue?

File: pkg/server/server.go [hunk 7]
```diff
@@ -341,37 +369,12 @@ func (s *Server) SendVerificationLink(c *gin.Context, unverifiedEmail string) {
...
-	err = generates.SendEmail(unverifiedEmail, "Email Verification", buf.String())
-	if err != nil {
-		log.Error("Error occurred while sending email for user: " + jwtUserCredentials.GetUID() + "error: " + err.Error())
-		c.JSON(http.StatusInternalServerError, gin.H{
-			"error": err.Error(),
-		})
-		return
-	}
-
 	c.JSON(http.StatusOK, gin.H{
```
Comment (avishnu) [inline pkg/server/server.go:378]: Trying to understand this line of code. How/where the http status is fetched from as after refactor, verification email is sent from inside the email manager?

File: pkg/server/server.go [hunk 8]
```diff
@@ -383,31 +386,32 @@ func (s *Server) VerifyEmail(c *gin.Context, redirectURL string) {
...
 		return
 	}
 	jwtUserCredentials := jwtUser.(*models.UserCredentials)
 
 	if jwtUserCredentials.UnverifiedEmail != nil {
-		*jwtUserCredentials.Email = jwtUserCredentials.GetUnverifiedEmail()
-		*jwtUserCredentials.UnverifiedEmail = ""
+		tmp := jwtUserCredentials.GetUnverifiedEmail()
+		jwtUserCredentials.Email = &tmp
+		jwtUserCredentials.UnverifiedEmail = nil
```
Comment (avishnu) [inline pkg/server/server.go:397]: What are we doing in the above 3 lines?

File: pkg/server/server.go [hunk 9]
```diff
@@ -383,31 +386,32 @@ func (s *Server) VerifyEmail(c *gin.Context, redirectURL string) {
...
 	} else {
 		log.Errorln("No email found to be verified for user uid: ", jwtUserCredentials.GetUID())
 		c.JSON(http.StatusBadRequest, gin.H{
 			"error": "No email found to be verified",
 		})
 		// Redirecting user to UI if no email found in field `unverified_email`
-		c.Redirect(http.StatusBadRequest, redirectURL)
+		c.Redirect(http.StatusPermanentRedirect, redirectURL)
```
Comment (avishnu) [inline pkg/server/server.go:404]: Why change the error code?

File: pkg/server/server.go [hunk 10]
```diff
@@ -383,31 +386,32 @@ func (s *Server) VerifyEmail(c *gin.Context, redirectURL string) {
...
 		return
 	}
 
 	_, err := usermanager.UpdateUserDetails(s.userStore, jwtUserCredentials)
 	if err != nil {
 		s.redirectError(c, err)
 		// Redirecting user to UI if updating the database fails
-		c.Redirect(http.StatusInternalServerError, redirectURL)
+		c.Redirect(http.StatusPermanentRedirect, redirectURL)
```
Comment (avishnu) [inline pkg/server/server.go:412]: Why change the error code?

File: pkg/server/server.go [hunk 11]
```diff
@@ -383,31 +386,32 @@ func (s *Server) VerifyEmail(c *gin.Context, redirectURL string) {
...
 	}
 
 	log.Infoln("Email: ", jwtUserCredentials.GetEmail(), " is verified successfully for user uid: ", jwtUserCredentials.GetUID())
-	c.Redirect(http.StatusFound, redirectURL)
+	c.Redirect(http.StatusPermanentRedirect, redirectURL)
```
Comment (avishnu) [inline pkg/server/server.go:416]: Why change the error code?

File: versionedController/v1/email/email.go [hunk 12]
```diff
@@ -54,7 +54,7 @@ func (emailController *EmailController) Get(c *gin.Context) {
 			"error": err.Error(),
 		})
 		// Redirecting user back to UI if access token is not valid
-		c.Redirect(http.StatusUnauthorized, redirectURL)
+		c.Redirect(http.StatusPermanentRedirect, redirectURL)
```
Comment (avishnu) [inline versionedController/v1/email/email.go:57]: Why change the error code?

File: versionedController/v1/signup/signup.go [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+	"net/http"
+
+	"github.com/gin-gonic/gin"
+	log "github.com/golang/glog"
+
+	"github.com/mayadata-io/kubera-auth/pkg/models"
+	controller "github.com/mayadata-io/kubera-auth/versionedController/v1"
+)
+
+//SignupController is the type the request in which the request will be parsed
```
Comment (avishnu) [inline versionedController/v1/signup/signup.go:13]: Grammatically incorrect statement.

File: versionedController/v1/signup/signup.go [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+func New() *SignupController {
+	return &SignupController{
+		routePath: controller.SignupRoute,
+	}
+}
+
+//Post registers a user in database taking minimal details needed
+func (signupController *SignupController) Post(c *gin.Context) {
+	signuplModel := &Model{}
+	err := c.BindJSON(signuplModel)
```
Comment (avishnu) [inline versionedController/v1/signup/signup.go:36]: lModel means?

File: manager/usermanager/createUser.go (hunk 1 above)
Comment (Sanjay1611) [reply manager/usermanager/createUser.go:33]: If `userRole` has been passed i.e. it is not empty then set it as the role of user to be created else just keep it as `user`

File: manager/usermanager/createUser.go (hunk 2 above)
Comment (Sanjay1611) [reply manager/usermanager/createUser.go:39]: This is to limit the fields that the new user is allowed to give as this function is being called from here too https://github.com/mayadata-io/kubera-auth/pull/45/files#diff-104a11712684e520f0affa95e6a053faeb713306a10f69ee99b9729cc9dd1996R272

File: manager/usermanager/createUser.go [hunk 15]
```diff
@@ -26,11 +28,28 @@ func CreateUser(userStore *store.UserStore, user *models.UserCredentials) (*mode
...
+		UID:             &uid,
+		UserName:        user.UserName,
+		Password:        &password,
+		Name:            user.Name,
+		UnverifiedEmail: user.UserName,
+		Kind:            models.LocalAuth,
```
[the remark this answers is not in the document]
Comment (Sanjay1611) [reply manager/usermanager/createUser.go:45]: Yes there is a seperate function for creating a user from oauth.

File: manager/usermanager/updateUser.go (hunk 3 above)
Comment (Sanjay1611) [reply manager/usermanager/updateUser.go:13]: Will update all the comments

File: manager/usermanager/updateUser.go (hunk 4 above)
Comment (Sanjay1611) [reply manager/usermanager/updateUser.go:19]: I guess it should be fine as its a pointer

File: pkg/models/user.go (hunk 6 above)
Comment (Sanjay1611) [reply pkg/models/user.go:84]: It will not cause any issue in the backend but it can cause some issue in UI as we will have to treat this case differently while signing up because rest all the users that will signup will surely contain this key

File: pkg/server/server.go (hunk 7 above)
Comment (Sanjay1611) [reply pkg/server/server.go:378]: Yes emailmanager is called here and if any error occurs it will return immediately. Else will reach this line of code and set status as `OK`

File: pkg/server/server.go (hunk 8 above)
Comment (Sanjay1611) [reply pkg/server/server.go:397]: Actually in the removed line number 392, we were  trying to set `Email` but its giving nil pointer exception as the `Email` can be empty.
So here we extract the `unverfied_email` (as we can't set it directly since its a pointer, dereferencing it afterwards will cause problems), and set it in the `email` field then making the `unverified_email` empty

File: pkg/server/server.go (hunk 9 above)
Comment (Sanjay1611) [reply pkg/server/server.go:404]: All the error codes are changed for redirection as for redirection only below error codes can be given else it will give error 
```
	StatusMultipleChoices   = 300 // RFC 7231, 6.4.1
	StatusMovedPermanently  = 301 // RFC 7231, 6.4.2
	StatusFound             = 302 // RFC 7231, 6.4.3
	StatusSeeOther          = 303 // RFC 7231, 6.4.4
	StatusNotModified       = 304 // RFC 7232, 4.1
	StatusUseProxy          = 305 // RFC 7231, 6.4.5
	_                       = 306 // RFC 7231, 6.4.6 (Unused)
	StatusTemporaryRedirect = 307 // RFC 7231, 6.4.7
	StatusPermanentRedirect = 308 // RFC 7538, 3

```

File: versionedController/v1/signup/signup.go (hunk 14 above)
Comment (Sanjay1611) [reply versionedController/v1/signup/signup.go:36]: It should be signupModel

File: pkg/models/user.go (hunk 6 above)
Comment (avishnu) [reply pkg/models/user.go:84]: But, admin does not "sign up", right?

File: manager/usermanager/updateUser.go (hunk 4 above)
Comment (avishnu) [reply manager/usermanager/updateUser.go:22]: Will execution hit the relevant switch case when the change happens within switch block?
Eg, will execution go from line no. 22 to line. 28?
Or from line 24 to line 35?

File: pkg/server/server.go [hunk 16]
```diff
@@ -383,31 +386,32 @@ func (s *Server) VerifyEmail(c *gin.Context, redirectURL string) {
 	if !exists {
 		s.redirectError(c, errors.ErrInvalidAccessToken)
 		// Redirecting user to UI if the user is not authorized
-		c.Redirect(http.StatusUnauthorized, redirectURL)
+		c.Redirect(http.StatusPermanentRedirect, redirectURL)
```
Comment (avishnu) [inline pkg/server/server.go:389]: What's the difference between s.redirect and c.Redirect? Why is it uppercase 'R' in c.Redirect and not lowercase?

File: pkg/server/server.go (hunk 8 above)
Comment (avishnu) [reply pkg/server/server.go:397]: Why do we need a string pointer instead of just a string? We don't need to worry about having nil, we can have as empty string.

File: pkg/server/server.go (hunk 9 above)
Comment (avishnu) [reply pkg/server/server.go:404]: How will UI know that an error has occurred if we are sending 301 as response code?

File: pkg/server/server.go (hunk 10 above)
Comment (avishnu) [reply pkg/server/server.go:412]: Same comment as https://github.com/mayadata-io/kubera-auth/pull/45/files#r592392465

File: pkg/server/server.go (hunk 11 above)
Comment (avishnu) [reply pkg/server/server.go:416]: Same comment as https://github.com/mayadata-io/kubera-auth/pull/45/files#r592392465

File: versionedController/v1/email/email.go (hunk 12 above)
Comment (avishnu) [reply versionedController/v1/email/email.go:57]: Same comment as https://github.com/mayadata-io/kubera-auth/pull/45/files#r592392465

File: manager/usermanager/updateUser.go (hunk 4 above)
Comment (Sanjay1611) [reply manager/usermanager/updateUser.go:22]: No @avishnu Only one of the switch case gets executed at a time in golang. It works as if breaak statement is hidden.

File: pkg/models/user.go (hunk 6 above)
Comment (Sanjay1611) [reply pkg/models/user.go:84]: Oh If the admin need not to signup then this is fine. Then we can remove this change. Normally we were making admin signup like a regular user.

File: pkg/server/server.go (hunk 16 above)
Comment (Sanjay1611) [reply pkg/server/server.go:389]: `c,Redirect` is a library function of `gin.Context` referred here as `c` and `s.redirect` is our function declared in this file. And `R` means function is exported and can be called by other files and `r` means function is not exported.

File: pkg/server/server.go (hunk 8 above)
Comment (Sanjay1611) [reply pkg/server/server.go:397]: OK we can change these types to string only. We can take this part in a separate PR.

File: pkg/server/server.go (hunk 9 above)
Comment (Sanjay1611) [reply pkg/server/server.go:404]: This status code will be shown to origin of the request that will be a button click on a email. Actual status and code that UI will come across is the above one with `c.JSON`

File: pkg/models/user.go (hunk 6 above)
Comment (avishnu) [reply pkg/models/user.go:84]: Aren't we creating the admin user during initialisation? Admin can only signin, not signup.
Is that not correct?

File: pkg/server/server.go (hunk 16 above)
Comment (avishnu) [reply pkg/server/server.go:389]: Thanks for the explanation of R vs r.
Trying to still understand why we need to call both s.redirect and c.Redirect.

**Comment by avishnu** [comment]
There is a conflict, please resolve.

**Review by vharsh — COMMENTED** [review]
PTAL, small issue re signup, also please add this endpoint to unauthenticatedLinks in router.go

File: pkg/models/user.go (hunk 6 above)
Comment (Sanjay1611) [reply pkg/models/user.go:84]: Yes but what we are doing previously is that we just create the admin credentials and assign his role as `admin` afterwards admin has to go through the same steps as a user i.e. verify his email and enter company details

File: manager/usermanager/createUser.go [hunk 17]
```diff
@@ -26,11 +28,28 @@ func CreateUser(userStore *store.UserStore, user *models.UserCredentials) (*mode
...
-	user.Password = &password
 	uid := uuid.Must(uuid.NewRandom()).String()
-	user.UID = &uid
-	err = userStore.Set(user)
-	return user.GetPublicInfo(), err
+	var role models.Role
+	if user.GetRole() != "" {
+		role = user.Role
+	} else {
+		role = models.RoleUser
```
[the remark this answers is not in the document]
Comment (vharsh) [reply manager/usermanager/createUser.go:36]: This comment is for line number 34 and not 36

File: pkg/server/server.go (hunk 16 above)
Comment (Sanjay1611) [reply pkg/server/server.go:389]: Okay I thinkwe are getting confused here with the wrong name implied to `s.redirect` function. What this function actually does is encodes an object or a data as a JSON into the context before giving the response. Naming is not correct here for `redirect` function.

**Comment by Sanjay1611** [comment]
> PTAL, small issue re signup, also please add this endpoint to unauthenticatedLinks in router.go

Have already added it in `router.go` @vharsh

File: manager/usermanager/createUser.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (Sanjay1611) [reply manager/usermanager/createUser.go:36]: This is a manager function. It is being called form 2 endpoints. One of them is SIgnup enpoint and the other is user creation endpoint by admin. 
If we see here https://github.com/mayadata-io/kubera-auth/pull/45/files#diff-1cbc60d38450f13ab57c0dafedcb6af3c687979beb08a644a069430d2fb56db0R20
there are limited fields that user can send for the signup. So role wouldn't be allowed here.

File: pkg/models/user.go (hunk 6 above)
Comment (avishnu) [reply pkg/models/user.go:84]: Admin need not verify or give company/role details as a mandatory case. I think we should mark the onboarding status as UnverifiedComplete during creation time (or in DB for already-created cases).

**Review by avishnu — APPROVED** [review]
Lgtm
